docs/api_design_v1beta2.md

```diff
@@ -0,0 +1,542 @@
+# AppMesh-K8s integration v1beta2 CRD API proposal
+
+## Introduction
+
+AppMesh is a service mesh product offered by Amazon. EKS team owns a controller that provides integration between AppMesh and Kubernetes using a [custom resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)(CRD) API.
+
+This document proposes a new version(v1beta2) of CRD API to address issues in current API design(v1beta1).
+
+## What is the problem
+
+There are several issues in the old CRD API model. Details in [AppMesh-K8s integration API issues](https://github.com/aws/aws-app-mesh-controller-for-k8s/issues/146). In summary:
+
+* The name of Kubernetes resources is tightly coupled with AppMesh names, which is inflexible and confusing.
+* The references between Kubernetes resources is ambiguous and relies on string hacks.
+* The security model for Kubernetes resources is not properly defined.
+* The VirtualNode CRD is implicitly coupled with Deployment resource.
+* The VirtualService cannot directly use VirtualNode as provider
+
+## Why is it changing
+
+We plan to announce k8s AppMesh controller as GA, and we should provide a stable and clean API for better customer experience.
+
+## Out of scope
+
+* The detailed implementation of new API is not discussed in this doc.
+
+## Roles:
+
+This document considers two user role:
+
+* **Cluster Administrator:** The cluster administrator (ops) is responsible for administration of entire clusters. They manage the lifecycle of AppMesh mesh and provide them to application developers.
+* **Application Developer**: The application developer (dev) is responsible for defining their application configuration and service composition.
+
+## Use cases:
+
+* As a Cluster Administrator for multi-tenant EKS cluster, I should be able to allocate a service mesh and control which users can create resources in this mesh.
+* As an Application Developer, I should be able to define the service routing logic and backend configuration for my application without Cluster Administrator's intervention.
+
+## Tenets:
+
+* The API should follow [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+* The API should be unambiguous, there should be only one obvious way for given task
+* The API should be explicit instead of rely on implicit logic
+* The API should be simple for 99% use case
+* The API should be open for future extensions
+
+## Design
+
+### API overview
```

Comment:
This is sort of a meta question. Have we thought at all about ergonomics and making it easier to meshify for the simple case? For example, the injector and controller act in unison, and we're able to automatically configure the mesh based off of the knowledge we have from the cluster. I know this is an implementation thing, so I suppose I'm wondering if the CRD design might act as a 1-way door to accomplishing that.

Reply:
> users only need to create virtualService/virtualRouters and virtualNodes

I'm thinking even being able to avoid specifying these in the first place, for the basic case of just enabling the mesh for the services you already have. We'd need extra software in the cluster but otherwise I think we can introspect everything we need? This seems to be how Linkerd works and I was impressed by how ergonomic it is to start if you're working within k8s.

That being said, taking another look at the diagram, it looks like there are no reverse references from the `Service` or `Deployment` to mesh resources, so I _think_ what you have is sufficient that we could down the road create a fancy discovery mechanism and all you have to do is specify that you want Envoy injected.

Reply:
@efe-selcuk I think there is no way to express AppMesh's functionality via the native k8s constructs. e.g. the GRPC routing or weighted Loadbalancing. The CRD approach is to fully expose appMesh's ability into k8s world.
And it's not a one way door. There can be a controller to "introspect" things in k8s and create these CRDs automatically if we feel the desire :D.

Reply:
Right right, we still need app mesh abstractions. But if someone already has a bunch of services talking to each other in the cluster (no fancy routing or weighted LB), that's a subset of what we provide. In my hypothetical world we'd generate the app mesh resources transparently and make it easy to export them for modification. But great, we're on the same page :+1:

Reply:
@efe-selcuk yeah, we can definitely do that for them in the future :D